integrations/pinecone/src/pinecone_haystack/document_store.py

```diff
@@ -0,0 +1,236 @@
+# SPDX-FileCopyrightText: 2023-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+import io
+import logging
+import os
+from typing import Any, Dict, List, Optional
+
+import pandas as pd
+import pinecone
+from haystack import default_from_dict, default_to_dict
+from haystack.dataclasses import Document
+from haystack.document_stores import DuplicatePolicy
+
+logger = logging.getLogger(__name__)
+
+# Pinecone has a limit of 1000 documents that can be returned in a query
+# with include_metadata=True or include_data=True
+# https://docs.pinecone.io/docs/limits
+TOP_K_LIMIT = 1_000
+
+
+class PineconeDocumentStore:
+    def __init__(
+        self,
+        *,
+        api_key: Optional[str] = None,
+        environment: str = "us-west1-gcp",
+        index: str = "default",
+        namespace: str = "default",
+        batch_size: int = 100,
+        dimension: int = 768,
+        **index_creation_kwargs,
+    ):
+        """
+        Creates a new PineconeDocumentStore instance.
+        It is meant to be connected to a Pinecone index and namespace.
+
+        :param api_key: The Pinecone API key. It can be explicitly provided or automatically read from the
+            environment variable PINECONE_API_KEY (recommended).
+        :param environment: The Pinecone environment to connect to. Defaults to "us-west1-gcp".
+        :param index: The Pinecone index to connect to. If the index does not exist, it will be created.
+            Defaults to "default".
+        :param namespace: The Pinecone namespace to connect to. If the namespace does not exist, it will be created
+            at the first write. Defaults to "default".
+        :param batch_size: The number of documents to write in a single batch. Defaults to 100, as recommended by
+            Pinecone.
+        :param dimension: The dimension of the embeddings. This parameter is only used when creating a new index.
+            Defaults to 768.
+        :param index_creation_kwargs: Additional keyword arguments to pass to the index creation method.
+            For example, you can specify `metric`, `pods`, `replicas`...
+            You can find the full list of supported arguments in the
+            [API reference](https://docs.pinecone.io/reference/create_index-1).
+
+        """
+        api_key = api_key or os.environ.get("PINECONE_API_KEY", None)
+        if not api_key:
+            msg = (
+                "PineconeDocumentStore expects a Pinecone API key. "
+                "Set the PINECONE_API_KEY environment variable (recommended) or pass it explicitly."
+            )
+            raise ValueError(msg)
+
+        pinecone.init(api_key=api_key, environment=environment)
+
+        if index not in pinecone.list_indexes():
```

Comment:
I would emit a log line in each case: we're using the existing index or we're creating one